src/worker/typosquat/types.rs

```diff
@@ -0,0 +1,238 @@
+            .limit(num)
+            .load_iter::<Crate, DefaultLoadingMode>(conn)?
+        {
+            let krate = result?;
+            crates.insert(
+                krate.id,
+                OwnedCrate {
+                    krate,
+                    owners: HashSet::new(),
+                },
+            );
+        }
+
+        // This query might require more low level knowledge of crate_owners than we really want
+        // outside of the models module. It would probably make more sense in the long term to have
+        // this live in the CrateOwner type, but for now I want to keep the typosquatting logic as
+        // self-contained as possible in case we decide not to go ahead with this in the longer
+        // term.
+        for result in crate_owners::table
+            .filter(crate_owners::deleted.eq(false))
+            .filter(crate_owners::crate_id.eq_any(crates.keys().collect::<Vec<_>>()))
+            .select((
+                crate_owners::crate_id,
+                crate_owners::owner_id,
+                crate_owners::owner_kind,
+            ))
+            .load_iter::<(i32, i32, i32), DefaultLoadingMode>(conn)?
+        {
+            let (crate_id, owner_id, owner_kind) = result?;
+            crates.entry(crate_id).and_modify(|owned| {
+                owned.owners.insert(UniqueOwner::new(owner_id, owner_kind));
+            });
+        }
+
+        Ok(Self {
+            crates: crates
+                .into_values()
+                .map(|owned| (owned.krate.name.clone(), owned))
+                .collect(),
+        })
+    }
+
+    /// Iterates over the names of the top crates.
+    pub(super) fn iter_names(&self) -> impl Iterator<Item = &str> {
+        self.crates.keys().map(String::as_str)
+    }
+}
+
+impl Corpus for TopCrates {
+    fn contains_name(&self, name: &str) -> typomania::Result<bool> {
+        Ok(self.crates.contains_key(name))
+    }
+
+    fn get(&self, name: &str) -> typomania::Result<Option<&dyn Package>> {
+        Ok(self.crates.get(name).map(|krate| krate as &dyn Package))
+    }
+
+    // We're using the default implementation of `possible_squat`, which will compare the owners of
+    // crates — a new crate won't be considered to be typosquatting a top crate if they share one
+    // or more owners. This prevents noise for things like an author releasing `foo2` for an
+    // existing `foo` crate.
+    //
+    // The one thing typogard-crates did that isn't implemented here is some basic NLP on the crate
+    // descriptions: this relies on the Python spaCy module, which we'd either have to reimplement
+    // the fuzzy matching logic from, or install a Python environment in the background-worker
+    // buildpack. Neither of those are particularly attractive options, so we're just going to
+    // accept a potentially higher level of false positives than we'd otherwise get.
+}
+
+/// A crate and its owners.
+pub(super) struct OwnedCrate {
+    krate: Crate,
```

Comment:
do we need the full `Crate` or would it be sufficient to only store the `description`? 

and since it isn't super-obvious, what is the description used for? I understand that in the original offline implementation NLP was used on the description for some checks, but we don't use that here, right? 🤔

Reply:
You are correct that actually keeping the description was pointless since we're not doing any fancy NLP stuff in the background job. Removed, and now we just keep the name and ownership information.